numpy/core/src/multiarray/number.c

```diff
@@ -353,23 +355,322 @@ PyArray_GenericInplaceUnaryFunction(PyArrayObject *m1, PyObject *op)
+            pos_ma_start = info.dli_fbase;
+            pos_ma_end = info.dli_fbase;
+        }
+        else {
+            init = -1;
+            return 0;
+        }
+        /* get c base address */
+        if (dladdr(&strcmp, &info)) {
+            pos_c_start = info.dli_fbase;
+            pos_c_end = info.dli_fbase;
+        }
+        else {
+            init = -1;
+            return 0;
+        }
+        init = 1;
+    }
+
+    for (i = 0; i < nptrs; i++) {
+        int in_python = 0;
+        int in_multiarray = 0;
+        int in_c = 0;
+#if NPY_ELIDE_DEBUG == 2
+        dladdr(buffer[i], &info);
+        printf("%s(%p) %s(%p)\n", info.dli_fname, info.dli_fbase,
+               info.dli_sname, info.dli_saddr);
+#endif
+        /* check stored boundaries first */
+        if (buffer[i] >= pos_python_start && buffer[i] <= pos_python_end) {
+            in_python = 1;
+        }
+        else if (buffer[i] >= pos_ma_start && buffer[i] <= pos_ma_end) {
+            in_multiarray = 1;
+        }
+        else if (buffer[i] >= pos_c_start && buffer[i] <= pos_c_end) {
+            in_c = 1;
+        }
+
+        /* new address, get base object */
+        if (!in_python && !in_multiarray && !in_c) {
+            if (dladdr(buffer[i], &info) == 0) {
+                init = -1;
+                ok = 0;
+                break;
+            }
+            /* new good function, store new end */
+            if (info.dli_fbase == pos_python_start) {
+                pos_python_end = NPY_NUMBER_MAX(buffer[i], pos_python_end);
+                in_python = 1;
+            }
+            else if (info.dli_fbase == pos_ma_start) {
+                pos_ma_end = NPY_NUMBER_MAX(buffer[i], pos_ma_end);
+                in_multiarray = 1;
+            }
+            else if (info.dli_fbase == pos_c_start) {
+                pos_c_end = NPY_NUMBER_MAX(buffer[i], pos_c_end);
+                in_c = 1;
+            }
+        }
+        /* have we left stack inside numpy? */
+        if (in_prolog_numpy && (!in_multiarray)) {
+            in_prolog_numpy = 0;
+        }
+        /* if outside initial numpy code only python and C functions are ok */
+        if (!in_prolog_numpy) {
+            if (!in_python && !in_c) {
+                ok = 0;
+                break;
+            }
+        }
+    }
+
+    /* all stacks after numpy are from python or C, we can elide */
+    if (ok) {
+        *cannot = 0;
+        return 1;
+    }
+    else {
+#if NPY_ELIDE_DEBUG != 0
+        puts("cannot elide due to c-api usage");
+#endif
+        *cannot = 1;
+        return 0;
+    }
+}
+
+/*
+ * check if m1 is a temporary (refcnt == 1) so we can do inplace operations
+ * instead of creating a new temporary
+ * cannot is set to true if it cannot be done even with swapped arguments
+ */
+static int
+can_elide_temp(PyArrayObject * m1, PyObject * m2, int * cannot)
+{
+    int i;
+    if (Py_REFCNT(m1) != 1 || !PyArray_CheckExact(m1) ||
```

Comment:
Py_REFCNT should be the NumPy specific PyArray_REFCOUNT

This way the whole function will not compile on PyPy, which is good as it is very CPython specific.